# Genotype maps

This folder contains all scripts used to generate the genotype maps (Extended Data Fig. 3–9, 11–13, 15).

## 0) Prepare the data for the analysis

This analysis requires the following input files:

- A GWAS file generated by GEMMA with four columns:
  (1) chromosome or scaffold name
  (2) scaffold:position
  (3) position
  (4) p-value  
  See `Example_GWAS_input.txt` for an example.

- A VCF file for the focal or ingroup species.

- A VCF file for the other closely related or comimetic species.

- A phenotype file for the focal or ingroup species with three columns:
  (1) sample ID
  (2) species
  (3) phenotype  
  See `Focal_species_information.txt`.

- A phenotype file for the other closely related or comimetic species with three columns:
  (1) sample ID
  (2) species
  (3) phenotype  
  See `Multisp_information.txt`.

- An order file indicating the display order of the focal or ingroup species followed by the other species in the genotype map.  
  See `order_file.txt`.

- A color file specifying the color associated with each phenotype.  
  See `Color.txt`.

## 1) Get pairwise relatedness between individuals

To control for  population structure, GEMMA computed the relatedness among samples. We used the following command to compute the pairwise relastedness matrix:

``` bash
$GEMMA -bfile $RESULTS_FOLDER -gk 1 > gemma_relatedness2.out
```

## 2) Run the GWAS

GEMMA was ran using the Univariate Linear Mixed Model. We employed the log-likelihood ratio test if the size effect of each SNP was significantly different from 0. 

``` bash
$GEMMA -bfile $RESULTS_FOLDER -lmm 4 -o assoc.gemma -k relatedness.cXX.txt > gemma.out
``` 

## 3) Plot the Manhattan plot

Manhattan plot summarising the results along the whole genome were plotted using R.

- In order to scaffolds plotted by increasing order we first ran the scaffold_size.py script on the reference genome.
``` bash
# Create file with scaffold sorted by ascending size
module load  Biopython/1.79-foss-2022a
python ./scaffold_size.py  $REFERENCE|sort -k 2 -nr|awk '{print $1}' > scaffold_order.txt
```

- The to get the number of SNPs used for the analyses (used to define the bonferroni threshold) we defined the THRESHOLD variable:
``` bash
THRESHOLD=$(cat $RESULTS/*assoc.gemma.log.txt|grep "analyzed SNPs/var" |awk '{print $7}')
``` 

- Finally the GWAS Manhattan plot is plotted using the R script Plot.R.
``` bash
# Make a png plot of the gwas 
Rscript ./Plot.R $RESULTS/*.assoc.gemma.assoc.txt $RESULTS/scaffold_order.txt $THRESHOLD $RESULTS
```


