<img src="Tiger wings.png" width="50%" height="50%"/>

# Genetic parallelism underpins convergent mimicry coloration across Lepidoptera

Welcome to the  GitHub repository for the scientific publication "***Genetic parallelism underpins convergent mimicry coloration across Lepidoptera***" by **Ben Chehida Y., van der Heijden S. M. E., Page J. E., et al. 2025**.

# Overview
This repository contains all the materials related to our research published in **TBA**. Here, you will find:

- Source custom scripts and pipelines used for data analysis

- Instructions to reproduce the results

- Additional explanations and documentation

# Purpose
Our work focuses on assessing the genomic basis of convergent evolution in mimetic butterflies and moths that diverged between 1 and 120 million years ago. Specifically, we aim to investigate whether similar wing colour patterns that have evolved repeatedly across these species arise from shared genetic mechanisms or through different evolutionary pathways. By studying multiple distantly related species exhibiting mimicry, this project seeks to deepen our understanding of how natural selection and genetic constraints influence the evolution of convergent phenotypes across recent vs deep evolutionary timescales.

# Contents

- **Reference_Genomes_assembly/** — Scripts used for genome assembly, generation, and curation.

- **BUSCO_TreePL/** — Scripts and related information for BUSCO phylogeny and molecular dating analyses, as presented in Figure 1.

- **Patternize/** — Scripts for quantifying and analyzing wing colour pattern variation across species.

- **Pop_Structure/** — Scripts for quantifying population genetic structure using PCA.

- **QTL/** — Scripts related to quantitative trait loci mapping to identify genomic regions associated with colour patterns in *H. pardalinus*.

- **GWAS/** — Scripts used to perform genome-wide association analyses presented in the paper.

- **GWAS_Peak_Alignments/** — Sliding window alignments and candidate regions identified in GWAS peak analyses between species, as shown in Figure 4.

- **RNAseq/** — Scripts and related information for: (1) generation of de novo transcriptomes, (2) annotation of *ivory*, and (3) differential expression analyses in *Mechanitis messenoides* and *Melinaea menophilus*.

- **Introgression/** — Scripts and related information for performing three types of introgression tests used in this study: (1) fd/fdm analyses, (2) Relate, and (3) sliding window NJ trees.

- **Genotype_maps/** - Pipeline and explanations to generate the genotype maps as showed in the Extended Data. 

- **Balancing_Selection/** — Scripts and related information for performing three types of balancing selection tests used in this study: (1) sliding window pi (across multiple species), (2) transpolymorphism analysis, and (3) MuteBass.

- **Transcription_Factor/** — Scripts and related information investigating transcription factor binding motifs around GWAS peaks.

- **Others/** — Additional scripts.
 
# Contact
For questions or feedback, please contact y.benchehida@sheffield.ac.uk.
