
# Principal component analyses and population structure

This folder contains all the scripts used to generate the Principal component analyses (PCA) scatter plots in the supplementary figures. The script PCA.sh was used to perform the PCA and the Rscript PCA_plot.R to plot the scatter plots for each species. 

## 1) Running the PCA
### A) LD prunning

The first step consist in prunning the data for SNPs under linkage disequilibrium. To create a linkage free dataset we used PLINK. More precisely using the "--indep-pairwise" flag, we looked at each window of 100 Kb and computed LD (r2) in steps of 10 bp each time. We prune any variant that show an r2 of greater than 0.1.

``` bash
$PLINK --vcf $VCF --double-id --allow-extra-chr --set-missing-var-ids @:# --indep-pairwise 100 10 0.1 
```

-> The optix PCA for M. marsaeus was performed without pruning. 

### B)  PERFORM THE PCA
- LD pruned dataset
  
``` bash
$PLINK --vcf $VCF --double-id --allow-extra-chr --set-missing-var-ids @:# --pca --extract *prune.in 
```


## 2) Identify common genes to all species

- For each one of the 18 species we extract all the protein sequences generated by BUSCO using the following command:

``` bash
for i in Bicyclus_anynana Chetone_histrio Heliconius_erato Heliconius_numata Hypothyris_anastasia Mechanitis_mazaeus Melinaea_isocomma Melinaea_menophilus Plutella_xylostella Biston_betularia Danaus_plexippus Heliconius_melpomene Heliconius_pardalinus Ithomia_salapia Mechanitis_messenoides Melinaea_marsaeus Melinaea_mothone Tithorea_tarricina; do cd $i; ls *faa > ../"$i"_busco_gene.txt ; cd .. ; done

paste *txt > common_busco_genes.txt
```

- Then the genes common to all the species were determined using the R script common_genes.R like this

``` bash
Rscript $SCRIPT/common_genes.R common_busco_genes.txt
```

- Finally we used the command below to keep the fasta protein sequences and nucleotide sequences common to the 18 species
``` bash
[ -e concatanated_genes_prot ] && rm -r concatanated_genes_prot
[ -e concatanated_genes_nucl ] && rm -r concatanated_genes_nucl
mkdir -p concatanated_genes_prot concatanated_genes_nucl

for i in  Bicyclus_anynana Chetone_histrio  Heliconius_erato Heliconius_numata Hypothyris_anastasia Mechanitis_mazaeus Melinaea_isocomma Melinaea_menophilus Plutella_xylostella Biston_betularia Danaus_plexippus Heliconius_melpomene Heliconius_pardalinus Ithomia_salapia Mechanitis_messenoides Melinaea_marsaeus Melinaea_mothone Tithorea_tarricina; do cat common_genes.txt|perl -pe 's/\.faa//g'| while read line; do cat $i/$line*faa >> concatanated_genes_prot/$line.faa; cat $i/$line*fna >> concatanated_genes_nucl/$line.fna; done; done
```

## 3) Use MUSCLE to align proteins and then pal2nal to reverse translate it to nucleotide
- For all the common protein, we used MUSCLE to align the protein sequences between the 18 species.
- Then we used the python script reorder_fasta.py to make sure that the order of the species in each fasta file (one for each common BUSCO gene) stay the same. This step is necesssary so later when we can concatanate all the genes together.
- Finally, we used the pal2nal software to reverse translate the aligned protein sequences -> nucleotide alignement.
  
``` bash
mkdir -p aligned_concatanated_genes_prot aligned_concatanated_genes_nucl
cd concatanated_genes_prot

for i in *faa; do 
	NAME=$(echo $i| perl -pe 's/faa/aln/g')
	NAME2=$(echo $i| perl -pe 's/faa/fna/g')
	cat $i |grep ">"|perl -pe 's/>//g' > order_"$NAME"
	$MUSCLE  -in $i -out ../aligned_concatanated_genes_prot/aligned_"$NAME"
	python2 $SCRIPT/reorder_fasta.py ../aligned_concatanated_genes_prot/aligned_"$NAME" order_"$NAME"  > ../aligned_concatanated_genes_prot/tmp_"$i"
	rm order_"$NAME" ../aligned_concatanated_genes_prot/aligned_"$NAME"
	mv ../aligned_concatanated_genes_prot/tmp_"$i" ../aligned_concatanated_genes_prot/aligned_"$NAME"
	$PAL2NAL ../aligned_concatanated_genes_prot/aligned_"$NAME" ../concatanated_genes_nucl/$NAME2 -output fasta -codontable 1 > ../aligned_concatanated_genes_nucl/aligned_"$NAME2"
done
```
