
# Principal component analyses and population structure

This folder contains all the scripts used to generate the Principal component analyses (PCA) scatter plots in the supplementary figures. The script PCA.sh was used to perform the PCA and the Rscript PCA_plot.R to plot the scatter plots for each species. 

## 1) Running the PCA
### A) LD prunning

The first step consist in prunning the data for SNPs under linkage disequilibrium. To create a linkage free dataset we used PLINK. More precisely using the "--indep-pairwise" flag, we looked at each window of 100 Kb and computed LD (r2) in steps of 10 bp each time. We prune any variant that shows an r2 of greater than 0.1.

``` bash
$PLINK --vcf $VCF --double-id --allow-extra-chr --set-missing-var-ids @:# --indep-pairwise 100 10 0.1 
```

-> The optix PCA for M. marsaeus was performed without pruning. 

### B)  PERFORM THE PCA
- PCA with LD pruning
  
``` bash
$PLINK --vcf $VCF --double-id --allow-extra-chr --set-missing-var-ids @:# --pca --extract *prune.in 
```

- PCA without LD pruning
``` bash
$PLINK --vcf $VCF --double-id --allow-extra-chr --set-missing-var-ids @:# --pca
```

## 2) Plotting the PCA scatterplots

- We plotted the PCA scatter plots (on PC1 and PC2) using the script "PCA_plot.R". This script runs like this:

``` bash
Rscript --vanilla ./PCA_plot.R *eigenvec *eigenval $PHENO $RESULTS_PATH $SPECIES
```
- *eigenvec *eigenval = eigen values and eigen vectors files generated by PLINK in step 1B.
- $PHENO = a phenotype file indicating to which group (species or phenotype) belong each sample in the VCf file.
- $SPECIES = the species of interest.
- $RESULTS_PATH = the path were the results should be stored.

- The scatter plots are generated in R using ggplot2

```
##############################
# Install required libraries #
##############################
library(plotly)
library(ggplot2)
library(Cairo)
options(bitmapType='cairo') # Necessary to make ggplotly works in the cluster
library(htmlwidgets)
library(rmarkdown)

###############
# Import data #
###############
# Eigenvectors
dat = commandArgs(trailingOnly=TRUE)
subset = read.table(dat[1])
subset = subset[,c(1,3,4)]
colnames(subset)=c("Ind","PC1","PC2")

# Eigenvalues
contribution = read.table(dat[2])
contribution = as.numeric(unlist(contribution))
  PC1_cont = (contribution[1]/sum(contribution))*100
  PC2_cont = (contribution[2]/sum(contribution))*100

# Phenotypes
pheno = read.table(dat[3])
subset = cbind(subset,pheno)
colnames(subset) = c("Ind","PC1","PC2","phenotype")
numb_color = length(unique(subset$phenotype))

###################
# Get output name #
###################
out_name = "Marsaeus"
print(paste(dat[4],"/",out_name,".html",sep=""))

########
# plot #
########
p1 <- ggplot(subset, aes(x=PC1, y=PC2,fill=phenotype)) + 
  geom_point(size = 3,color="black",shape=21,aes(text = Ind)) + 
  theme_bw() + 
  theme(plot.title = element_text(hjust = 0.5)) +
  scale_fill_manual(values=rainbow(numb_color)) +
  theme(legend.margin=margin(0,0,0,0),
        legend.box.margin=margin(-10,20,-5,0)) +
    xlab(paste("PC1 (",round(x=PC1_cont,2),"%)",sep="")) +
  ylab(paste("PC2 (",round(x=PC2_cont,2),"%)",sep=""))  

pdf(paste(dat[4],"/",out_name,".pdf",sep=""),12,6)
p1
dev.off()
```
