# BUSCO phylogenetic tree and treePL dating 

This folder contains all the scripts used to generate the dated BUSCO phylogenetic tree (Figure 1).

## 1) Running BUSCO on each reference genome

The first step (busco.sh) involves finding and retrieving all the BUSCO genes/protein sequences for each of the 18 reference genomes used (hereafter referred to as $ref_genome). We ran the BUSCO software using default option:

``` bash
busco -m genome -i $ref_genome -l lepidoptera_odb10 -c 1 --config config.ini
```

## 2) Identify common genes to all species

- For each one of the 18 species we extract all the protein sequences generated by BUSCO using the following command:

``` bash
for i in Bicyclus_anynana Chetone_histrio Heliconius_erato Heliconius_numata Hypothyris_anastasia Mechanitis_mazaeus Melinaea_isocomma Melinaea_menophilus Plutella_xylostella Biston_betularia Danaus_plexippus Heliconius_melpomene Heliconius_pardalinus Ithomia_salapia Mechanitis_messenoides Melinaea_marsaeus Melinaea_mothone Tithorea_tarricina; do cd $i; ls *faa > ../"$i"_busco_gene.txt ; cd .. ; done

paste *txt > common_busco_genes.txt
```

- Then the genes common to all the species were determined using the R script common_genes.R like this

``` bash
Rscript $SCRIPT/common_genes.R common_busco_genes.txt
```

- Finally we used the command below to keep the fasta protein sequences and nucleotide sequences common to the 18 species
``` bash
[ -e concatanated_genes_prot ] && rm -r concatanated_genes_prot
[ -e concatanated_genes_nucl ] && rm -r concatanated_genes_nucl
mkdir -p concatanated_genes_prot concatanated_genes_nucl

for i in  Bicyclus_anynana Chetone_histrio  Heliconius_erato Heliconius_numata Hypothyris_anastasia Mechanitis_mazaeus Melinaea_isocomma Melinaea_menophilus Plutella_xylostella Biston_betularia Danaus_plexippus Heliconius_melpomene Heliconius_pardalinus Ithomia_salapia Mechanitis_messenoides Melinaea_marsaeus Melinaea_mothone Tithorea_tarricina; do cat common_genes.txt|perl -pe 's/\.faa//g'| while read line; do cat $i/$line*faa >> concatanated_genes_prot/$line.faa; cat $i/$line*fna >> concatanated_genes_nucl/$line.fna; done; done
```

## 3) Use MUSCLE to align proteins and then pal2nal to reverse translate it to nucleotide
- For all the common protein, we used MUSCLE to align the protein sequences between the 18 species.
- Then we used the python script reorder_fasta.py to make sure that the order of the species in each fasta file (one for each common BUSCO gene) stay the same. This step is necesssary so later when we can concatanate all the genes together.
- Finally, we used the pal2nal software to reverse translate the aligned protein sequences -> nucleotide alignement.
  
``` bash
mkdir -p aligned_concatanated_genes_prot aligned_concatanated_genes_nucl
cd concatanated_genes_prot

for i in *faa; do 
	NAME=$(echo $i| perl -pe 's/faa/aln/g')
	NAME2=$(echo $i| perl -pe 's/faa/fna/g')
	cat $i |grep ">"|perl -pe 's/>//g' > order_"$NAME"
	$MUSCLE  -in $i -out ../aligned_concatanated_genes_prot/aligned_"$NAME"
	python2 $SCRIPT/reorder_fasta.py ../aligned_concatanated_genes_prot/aligned_"$NAME" order_"$NAME"  > ../aligned_concatanated_genes_prot/tmp_"$i"
	rm order_"$NAME" ../aligned_concatanated_genes_prot/aligned_"$NAME"
	mv ../aligned_concatanated_genes_prot/tmp_"$i" ../aligned_concatanated_genes_prot/aligned_"$NAME"
	$PAL2NAL ../aligned_concatanated_genes_prot/aligned_"$NAME" ../concatanated_genes_nucl/$NAME2 -output fasta -codontable 1 > ../aligned_concatanated_genes_nucl/aligned_"$NAME2"
done
```

## 4) Remove aligned genes with an excessive number of "-"
We excluded poorly aligned genes (i.e. genes with too many "-" in the alignment) by keeping only genes with at most 5% of "-". We used the indels_count.py script to count the number "-" in the alignement. indels_count.py relies on Biopython. The command below was used to remove aligments with too many "-":

``` bash
cd ../aligned_concatanated_genes_nucl

for i in *fna; do 
	INDELS=$(python -W ignore $SCRIPT/indels_count.py ../aligned_concatanated_genes_nucl/$i); 
	if [[ "$INDELS" -gt 5 ]]; then 
		echo $i; echo $INDELS; rm ../aligned_concatanated_genes_nucl/$i; 
	fi; 
done
```
  
## 5) Replace fasta headers with only the species name
This step is essential so in step 6 we can merge together all the gene sequence aligments in a single superaligment.  

``` bash
cd ../aligned_concatanated_genes_nucl
for j in *.fna; do 
	for i in Bicyclus_anynana Chetone_histrio  Heliconius_erato Heliconius_numata Hypothyris_anastasia Mechanitis_mazaeus Melinaea_isocomma Melinaea_menophilus Plutella_xylostella Biston_betularia Danaus_plexippus Heliconius_melpomene Heliconius_pardalinus Ithomia_salapia Mechanitis_messenoides Melinaea_marsaeus Melinaea_mothone Tithorea_tarricina; do 
		sed -i -E ':a;N;$!ba; s/(>)([a-zA-Z0-9\_\.]+):([0-9]+)-([0-9]+)/\1'"${i}"'/1' $j; 
	done; 
done
```

## 6) Concatenate all the genes in a single fasta file
We concatenated all the  gene sequence aligments in a single superaligment fasta file. This step rely on the softwawre SeqKit.

``` bash
$SEQKIT concat *.fna  > ../final.fna
```

## 7)Maximum Likelihood tree of the busco genes

- The relationship between the busco genes was inferred using RAxML v8 using the GTRGAMMAI substitution model (RaxML.sh). 
``` bash
$RAXML -s final.fna -m GTRGAMMAI -n BUSCO_PHYLO -p 1234
```

- The following command was used to remove information on branches length (which prevent later steps to work).
  
``` bash
sed -r 's/[0-9:.]//g' input.tre > output_topo.tre
```

NB: Steps 8 to 12 are largely inspired by this: [GitHub Pages](https://github.com/sunray1/treepl/tree/master)

## 8) Get bootstrap alignments
The following command was used to bootstrap the alignment fasta file (final.fna). We generated 100 bootstrap alignments using the option -f j in RAxML.
``` bash
$RAXML -f j -m GTRGAMMAI -s final.fna -n BS -# 100 -b $RANDOM
```

## 9) Getting a tree for each bootstrapped alignment

- We optimized the model parameters and branch lengths of the bootstrapped alignments while constraining the topology to match the previously inferred ML tree (step 7) using the -f e option in RAxML.

``` bash
for i in {0..99};
do
	$RAXML -f e -t output_topo.tre -m GTRGAMMAI -s final.fna.BS"$i" -n BS_"$i";
done
```

- We combine the tree inferred for each alignement in a single file using the following command
``` bash
cat RAxML_result.BS* > RAxML_bootstrap.bootstrap_all.tre
```

## 10) Root the bootstrap trees 
- We rooted all the bootstrap trees using Plutella xylostella as an outgroup with the program phyx. 

``` bash
/shared/biology/bioldata1/bl-kd684/yacine/Conv_Evol/Tools/phyx/src -t RAxML_bootstrap.bootstrap_all.tre -g Plutella_xylostella -o RAxML_bootstrap.bootstrap_all_rooted.tre
```
- Create a separate file for each bootstrap tree so later we can run TreePL on each tree 

```
mkdir -p 100_trees
split -l 1 RAxML_bootstrap.bootstrap_all_rooted.tre 100_trees/RAxML_bestTree.BSone_tree
for i in 100_trees/*; do mv $i $i.tre; done
```

## 11) Running TreePL

Divergence time estimates were obtained using a penalized-likelihood based approach implemented in TreePL (script submit_treepl.sh).
The script consist in 3 steps:

- 1) A priming step erformed on each of the 100 bootstrap trees

``` bash
mkdir -p ../Results/step1_output
for i in {1..100}; do python3 ./run_treepl.py ../Results/100_trees $i ../Results/step1_output; done
```

- 2) A cross-validation step that we run 10 times for each 100 bootstrap trees

```
for run in {1..10}; \
do
	for i in {1..100};
	do
		./run_treepl2.py ../Results/100_trees $i ../Results/step2."$run"_output;
	done 
done
```

- 3) A dating step based on the best smoothing parameters

```
for i in {1..100}; do python3 ./run_treepl3.py ../Results/100_trees $i ../Results/step3_output; done
```

## 12) Summarise the divergence time estimates using treeAnnotator

We used the TreeAnnotator utilities from the Beast package to calculated the 95% highest posterior density for the node ages using a burnin of 10%. 

```
cd ../Results/step3_output
cat *.tre > treeall_dated.tre
treeannotator -burnin 10 treeall_dated.tre out.txt
```

## 13) Plotting the results with 
